File: src/models/tools/image_generator.py
from typing import Dict, ClassVar, Literal
from pydantic import BaseModel, Field


class ImageGeneratorToolConfig(BaseModel):
    api_key: str | None = Field(
        default=None,
        description="The Stability AI API key to use for the image generator.",
    )


class ImageGeneratorTool(BaseModel):
    config_id: ClassVar[str] = "img_gen"
    name: Literal["image_generator"] = "image_generator"
    tool_type: Literal["local"] = "local"
    label: Literal["Image Generator"] = "Image Generator"
    description: str = (
        "Generate images from the provided text using the Stability AI API"
    )
    custom_prompt: ClassVar[
        str
    ] = """- Provide a text prompt to generate an image.
- Show the image to the user by using the absolute link to the image from the tool output."""
    config: ImageGeneratorToolConfig | None = Field(
        default=ImageGeneratorToolConfig(),
    )
    enabled: bool = False

File: src/ragapp/admin-ui/client/tools/image_generator.ts
import { z } from "zod";

export const ImageGeneratorToolConfig = z.object({
  name: z.literal("img_gen"),
  label: z.string().nullable().optional(),
  enabled: z.boolean().nullable().optional(),
  description: z.string(),
  priority: z.number(),
  config: z
    .object({
      api_key: z
        .string()
        .nullable()
        .refine(
          (data) => {
            return data && data.length > 0;
          },
          { message: "API Key is required" },
        ),
    })
    .nullable()
    .optional(),
});

export type ImageGeneratorToolConfigType = z.infer<
  typeof ImageGeneratorToolConfig
>;

export const DEFAULT_IMAGE_GENERATOR_TOOL_CONFIG: ImageGeneratorToolConfigType =
  {
    name: "img_gen",
    label: "Image Generator",
    description:
      "Generate images from the provided text using the Stability AI API",
    priority: 0,
    config: {
      api_key: "",
    },
    enabled: false,
  };
